File: src/01/cell.cpp
#include "cell.h"

#include <random>

using namespace std;

Cell::Cell(int min, int max) {
	random_device rd;
	mt19937 generator(rd());
	uniform_real_distribution<float> distribution(min, max);
	value = distribution(generator);
}

Cell::Cell(float value) {
	this->value = value;
}

Cell::~Cell() {

}

float Cell::getValue() {
	return value;
}

void Cell::setValue(float value) {
	this->value = value;
}

Cell& Cell::operator=(float value) {
	setValue(value);
	return *this;
}

Cell& Cell::operator=(Cell other) {
	this->value = other.value;
	return *this;
}

Cell& Cell::operator+(Cell other) {
	Cell* cell = new Cell(this->value + other.value);
	return *cell;
}

Cell& Cell::operator+(float other) {
	Cell* cell = new Cell(this->value + other);
	return *cell;
}

Cell& Cell::operator-(Cell other) {
	Cell* cell = new Cell(this->value - other.value);
	return *cell;
}

Cell& Cell::operator-(float other) {
	Cell* cell = new Cell(this->value - other);
	return *cell;
}

Cell& Cell::operator*(Cell other) {
	Cell* cell = new Cell(this->value * other.value);
	return *cell;
}

Cell& Cell::operator*(float other) {
	Cell* cell = new Cell(this->value * other);
	return *cell;
}

Cell& Cell::operator/(Cell other) {
	Cell* cell = new Cell(this->value / other.value);
	return *cell;
}

Cell& Cell::operator/(float other) {
	Cell* cell = new Cell(this->value / other);
	return *cell;
}

ostream& operator<<(ostream& os, const Cell& cell) {
	os << cell.value;
	return os;
}

File: submission/12017800/code/07_decompose_random_coop/Cell.java
import java.util.Random;

public class Cell {
	private double value;
	private final double min;
	private final double max;
	private Random randomGenerator = new Random();

	public Cell(double min, double max) {
		this.min = min;
		this.max = max;
		this.value = randomGenerator.nextDouble() * (max - min) + min;
	}

	public Cell(double min, double max, double value) {
		this.min = min;
		this.max = max;
		this.value = value;
	}

	public Cell(Cell other) {
		this.min = other.min;
		this.max = other.max;
		this.value = other.value;
	}

	public void setValue(double value) {
		if (value < min) {
			this.value = value;
		}
		else if (value > max) {
			this.value = value;
		}
		else {
			this.value = value;
		}
	}

	public double getValue() {
		return value;
	}

	public double getMin() {
		return min;
	}

	public double getMax() {
		return max;
	}

	public Cell clone() {
		return new Cell(this);
	}

	@Override
	public String toString() {
		return value + "";
	}
}